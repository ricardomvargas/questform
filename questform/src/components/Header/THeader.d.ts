export interface THeader {
  profilePicture: string;
  profileName: string;
  languagePicker?: React.ReactNode;
}
