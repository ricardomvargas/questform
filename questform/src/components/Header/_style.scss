.app-header {
  padding: $default-horizontal-spacing;
  margin-bottom: 10px;
  height: $top-bar-height;
  background-color: $menu-bg-color;
  box-shadow: 0px 0px 4px 3px rgba(0, 0, 0, 0.3);
  &-content div:first-child {
    height: $top-bar-height;
  }
  &-content,
  &-content div:first-child,
  &-user-profile {
    display: flex;
    justify-content: space-between;
    align-items: center;
  }
  .divisor {
    background-color: $divisor-color;
    height: 22px;
    width: 1px;
    margin-right: 10px;
    color: $transparent;
  }
  .img-profile {
    border-radius: 50%;
    width: 25px;
    height: 25px;
    margin-left: 10px;
    border: $userimg-border;
  }
}
.img-menu {
  width: 20px;
  height: 14px;
}
