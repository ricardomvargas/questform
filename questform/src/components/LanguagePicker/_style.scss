%menu-item-options-default {
  background-color: $blue-white;
  border: solid 2px $blue-gray;
  border-radius: 20px;
  position: absolute;
  width: 45px;
  padding: 2px;
  top: 32px;
}
.lang-menu {
  .img-flag {
    width: 20px;
    height: 20px;
  }
  .img-options {
    width: 12px;
    height: 12px;
    margin-left: 3px;
  }
  .display-langs {
    border: none;
    background: none;
    padding: 0;
    margin: 0;
  }
  &-options {
    @extend %menu-item-options-default;
    display: flex;
    justify-content: space-between;
  }
  &-options-hide {
    @extend %menu-item-options-default;
    display: none;
  }
}
