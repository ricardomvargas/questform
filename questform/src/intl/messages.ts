import appMessagesEN from './app/en';
import appMessagesPT from './app/pt';
import dashboardEN from './pages/dashboard/en';
import dashboardPT from './pages/dashboard/pt';
import questionEN from './pages/question/en';
import questionPT from './pages/question/pt';
import surveyEN from './pages/survey/en';
import surveyPT from './pages/survey/pt';

export default {
  en: {
    ...appMessagesEN,
    ...dashboardEN,
    ...questionEN,
    ...surveyEN,
  },
  pt: {
    ...appMessagesPT,
    ...dashboardPT,
    ...questionPT,
    ...surveyPT,
  },
};
