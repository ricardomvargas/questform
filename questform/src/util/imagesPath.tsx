export const menuIcon = '/assets/images/menu-icon.png';
export const brFlagRounded = '/assets/images/br-flag-rounded.png';
export const ukFlagRounded = '/assets/images/uk-flag-rounded.png';
export const arrowDownImg = '/assets/images/arrow-down.png';
export const closeImg = '/assets/images/close.png';
export const arrowRight = '../../assets/images/arrow-right.png';
export const profileImg = './assets/images/me.jpg';
