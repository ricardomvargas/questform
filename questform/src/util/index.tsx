export const getStatus = (codigo: number) => {
  switch (codigo) {
    case 1:
      return 'offline';
    case 2:
      return 'online';
    case 3:
      return 'publishing';
    case 4:
      return 'finished';
    case 5:
      return 'canceled';
    case 6:
      return 'archived';
  }
};

export const isInt = (n: number) => n % 1 === 0;

export const getIntegerDivision = (value: number, divider: number) => {
  const result = value / divider;
  if (isInt(result)) return result;
  const newResult = parseInt(result.toString());
  return newResult + 1;
};
