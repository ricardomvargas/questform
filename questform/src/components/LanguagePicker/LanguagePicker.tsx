import React, { useState, useContext } from 'react';
import { useIntl } from 'react-intl';

import LangContext from '../../context';

import brFlagRounded from '../../assets/images/br-flag-rounded.png';
import ukFlagRounded from '../../assets/images/uk-flag-rounded.png';
import arrowDownImg from '../../assets/images/arrow-down.png';
import closeImg from '../../assets/images/close.png';

import { storeLanguage, appLangs } from '../../intl/languageStorage';

const LanguagePicker = () => {
  const intl = useIntl();
  const { updateAppLang, appLang } = useContext(LangContext);
  const [langOptionsVisible, setLangOptionsVisible] = useState(false);
  const [langOptionsLeftPosition, setlangOptionsLeftPosition] = useState<number>();

  const getButtonFlagPosition = (el: HTMLImageElement) => {
    let x: number = 0;
    while (el && !isNaN(el.offsetLeft) && !isNaN(el.offsetTop)) {
      x += el.offsetLeft - el.scrollLeft;
      el = el.offsetParent as HTMLImageElement;
    }
    return x - 36;
  };

  const updateMenuLangVisibility = (): void => {
    setlangOptionsLeftPosition(
      getButtonFlagPosition(document.getElementById('btn-show-menu-item') as HTMLImageElement)
    );
    setLangOptionsVisible(!langOptionsVisible);
  };

  const handleLanguageChange = (newLang: string): void => {
    storeLanguage(newLang);
    if (updateAppLang) updateAppLang(newLang);
    updateMenuLangVisibility();
  };

  const closeMenuItemOptions = (): void => setLangOptionsVisible(false);
  // #TODO: Analisar se não é melhor usar um useEffect para registrar esse evento
  window.addEventListener('resize', closeMenuItemOptions);

  return (
    <section className="lang-menu">
      <div className="lang-menu-item">
        <img
          src={appLang === appLangs.english ? ukFlagRounded : brFlagRounded}
          className="img-flag"
          alt={appLang}
          title={appLang}
          onClick={updateMenuLangVisibility}
        />
        <img
          id="btn-show-menu-item"
          className="img-options"
          src={langOptionsVisible ? closeImg : arrowDownImg}
          alt={
            langOptionsVisible
              ? intl.formatMessage({ id: 'buttonClose' })
              : intl.formatMessage({ id: 'buttonOpen' })
          }
          title={
            langOptionsVisible
              ? intl.formatMessage({ id: 'buttonClose' })
              : intl.formatMessage({ id: 'buttonOpen' })
          }
          onClick={updateMenuLangVisibility}
        />
      </div>
      <div
        style={{ left: langOptionsLeftPosition }}
        className={langOptionsVisible ? 'lang-menu-options' : 'lang-menu-options-hide'}
      >
        <img
          src={ukFlagRounded}
          className="img-flag"
          alt={intl.formatMessage({ id: 'changeToEnglish' })}
          title={intl.formatMessage({ id: 'changeToEnglish' })}
          onClick={() => handleLanguageChange(appLangs.english)}
        />
        <img
          src={brFlagRounded}
          className="img-flag"
          alt={intl.formatMessage({ id: 'changeToPortuguese' })}
          title={intl.formatMessage({ id: 'changeToPortuguese' })}
          onClick={() => handleLanguageChange(appLangs.portugueseBR)}
        />
      </div>
    </section>
  );
};

export default LanguagePicker;
