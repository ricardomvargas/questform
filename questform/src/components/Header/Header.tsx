import React, { useState } from 'react';
import { useIntl } from 'react-intl';

import { THeader } from './THeader';
import { AppLink } from '../Menu/TMenu';

import Menu from '../Menu/Menu';
import { menuIcon } from '../../util/imagesPath';

import { dashboardRoute, surveysListRoute, questionsListRoute } from '../../util/routes';

const Header = ({ profilePicture, profileName, languagePicker }: THeader) => {
  const [menuVisible, setMenuVisible] = useState(false);
  const intl = useIntl();

  const appLinks: AppLink[] = [
    {
      title: intl.formatMessage({ id: 'menuHome' }),
      imgId: 'imgGo',
      url: dashboardRoute,
    },
    {
      title: intl.formatMessage({ id: 'menuSurveys' }),
      imgId: 'imgGo',
      url: surveysListRoute,
    },
    {
      title: intl.formatMessage({ id: 'menuQuestions' }),
      imgId: 'imgGo',
      url: questionsListRoute,
    },
  ];

  const handleMenuClick = (): void => setMenuVisible(!menuVisible);

  return (
    <header>
      <section className="app-header">
        <div className="app-header-content">
          <div>
            <img
              src={menuIcon}
              style={{ display: menuVisible ? 'none' : 'inline-block' }}
              className="img-menu"
              alt={intl.formatMessage({ id: 'menu' })}
              title={intl.formatMessage({ id: 'menu' })}
              onClick={handleMenuClick}
            />
          </div>
          <div className="app-header-user-profile">
            <span className="divisor"></span>
            {languagePicker}
            <img
              src={profilePicture}
              className="img-profile"
              alt={intl.formatMessage({ id: 'profile' })}
              title={profileName}
            />
          </div>
        </div>
      </section>
      <Menu
        menuIcon={menuIcon}
        menuVisible={menuVisible}
        onMenuClick={handleMenuClick}
        appLinks={appLinks}
      />
    </header>
  );
};

export default Header;
