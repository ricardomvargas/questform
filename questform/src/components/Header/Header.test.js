/**
 * @jest-environment jsdom
 */
import React from 'react';
import '@testing-library/jest-dom';
import { IntlProvider } from 'react-intl';
import { render } from '@testing-library/react';

import LangContext from '../../context';

import Header from './Header';
import LanguagePicker from '../LanguagePicker/LanguagePicker';

import messages from '../../intl/messages';
import { profileImg } from '../../util/imagesPath';

const updateAppLang = jest.fn();
const appLang = 'en';

const Wraper = ({ children }) => (
  <IntlProvider locale={appLang} messages={messages.en}>
    {children}
  </IntlProvider>
);

test('Check if match with snapshot', () => {
  const header = render(
    <Wraper>
      <Header
        profilePicture={profileImg}
        profileName={'Ricardo Vargas'}
        languagePicker={
          <LangContext.Provider value={{ updateAppLang, appLang }}>
            <LanguagePicker />
          </LangContext.Provider>
        }
      />
    </Wraper>
  );
  expect(header).toMatchSnapshot();
});
