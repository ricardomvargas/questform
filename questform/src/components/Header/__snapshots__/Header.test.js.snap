// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if match with snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header>
        <section
          class="app-header"
        >
          <div
            class="app-header-content"
          >
            <div>
              <img
                alt="Menu"
                class="img-menu"
                src="/assets/images/menu-icon.png"
                style="display: inline-block;"
                title="Menu"
              />
            </div>
            <div
              class="app-header-user-profile"
            >
              <span
                class="divisor"
              />
              <section
                class="lang-menu"
              >
                <div
                  class="lang-menu-item"
                >
                  <img
                    alt="en"
                    class="img-flag"
                    src="/assets/images/uk-flag-rounded.png"
                    title="en"
                  />
                  <img
                    alt="Open"
                    class="img-options"
                    id="btn-show-menu-item"
                    src="/assets/images/arrow-down.png"
                    title="Open"
                  />
                </div>
                <div
                  class="lang-menu-options-hide"
                >
                  <img
                    alt="Change to english"
                    class="img-flag"
                    src="/assets/images/uk-flag-rounded.png"
                    title="Change to english"
                  />
                  <img
                    alt="Change to portuguese"
                    class="img-flag"
                    src="/assets/images/br-flag-rounded.png"
                    title="Change to portuguese"
                  />
                </div>
              </section>
              <img
                alt="My profile"
                class="img-profile"
                src="./assets/images/me.jpg"
                title="Ricardo Vargas"
              />
            </div>
          </div>
        </section>
        <nav
          class="menu"
          style="width: 0px;"
        >
          <section
            class="menu-top"
          >
            <div>
              <img
                alt="Menu"
                class="img-menu"
                src="/assets/images/menu-icon.png"
                title="Menu"
              />
            </div>
          </section>
          <section>
            <ul
              class="menu-links"
            />
          </section>
        </nav>
      </header>
    </div>
  </body>,
  "container": <div>
    <header>
      <section
        class="app-header"
      >
        <div
          class="app-header-content"
        >
          <div>
            <img
              alt="Menu"
              class="img-menu"
              src="/assets/images/menu-icon.png"
              style="display: inline-block;"
              title="Menu"
            />
          </div>
          <div
            class="app-header-user-profile"
          >
            <span
              class="divisor"
            />
            <section
              class="lang-menu"
            >
              <div
                class="lang-menu-item"
              >
                <img
                  alt="en"
                  class="img-flag"
                  src="/assets/images/uk-flag-rounded.png"
                  title="en"
                />
                <img
                  alt="Open"
                  class="img-options"
                  id="btn-show-menu-item"
                  src="/assets/images/arrow-down.png"
                  title="Open"
                />
              </div>
              <div
                class="lang-menu-options-hide"
              >
                <img
                  alt="Change to english"
                  class="img-flag"
                  src="/assets/images/uk-flag-rounded.png"
                  title="Change to english"
                />
                <img
                  alt="Change to portuguese"
                  class="img-flag"
                  src="/assets/images/br-flag-rounded.png"
                  title="Change to portuguese"
                />
              </div>
            </section>
            <img
              alt="My profile"
              class="img-profile"
              src="./assets/images/me.jpg"
              title="Ricardo Vargas"
            />
          </div>
        </div>
      </section>
      <nav
        class="menu"
        style="width: 0px;"
      >
        <section
          class="menu-top"
        >
          <div>
            <img
              alt="Menu"
              class="img-menu"
              src="/assets/images/menu-icon.png"
              title="Menu"
            />
          </div>
        </section>
        <section>
          <ul
            class="menu-links"
          />
        </section>
      </nav>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
